system-contracts/contracts/BatchAggregator.sol

```diff
@@ -0,0 +1,354 @@
+        );
+        calldataPtr += 4 + numberOfMessages * 4;
+
+        /// Check bytecodes
+        uint32 numberOfBytecodes = uint32(bytes4(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 4]));
+        uint256 bytecodeSliceStart = calldataPtr;
+        calldataPtr += 4;
+        bytes32 reconstructedChainedL1BytecodesRevealDataHash;
+        for (uint256 i = 0; i < numberOfBytecodes; ++i) {
+            uint32 currentBytecodeLength = uint32(
+                bytes4(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 4])
+            );
+            calldataPtr += 4 + currentBytecodeLength;
+        }
+
+        bytecodeStorage[chainId].push(_totalL2ToL1PubdataAndStateDiffs[bytecodeSliceStart:calldataPtr]);
+        /// Check State Diffs
+        /// encoding is as follows:
+        /// header (1 byte version, 3 bytes total len of compressed, 1 byte enumeration index size)
+        /// body (`compressedStateDiffSize` bytes, 4 bytes number of state diffs, `numberOfStateDiffs` * `STATE_DIFF_ENTRY_SIZE` bytes for the uncompressed state diffs)
+        /// encoded state diffs: [20bytes address][32bytes key][32bytes derived key][8bytes enum index][32bytes initial value][32bytes final value]
+        require(
+            uint256(uint8(bytes1(_totalL2ToL1PubdataAndStateDiffs[calldataPtr]))) ==
+                STATE_DIFF_COMPRESSION_VERSION_NUMBER,
+            "state diff compression version mismatch"
+        );
+        uint256 stateDiffSliceStart = calldataPtr;
+        calldataPtr++;
+
+        uint24 compressedStateDiffSize = uint24(bytes3(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 3]));
+        calldataPtr += 3;
+
+        uint8 enumerationIndexSize = uint8(bytes1(_totalL2ToL1PubdataAndStateDiffs[calldataPtr]));
+        calldataPtr++;
+
+        bytes calldata compressedStateDiffs = _totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr +
+            compressedStateDiffSize];
+        calldataPtr += compressedStateDiffSize;
+
+        bytes calldata totalL2ToL1Pubdata = _totalL2ToL1PubdataAndStateDiffs[:calldataPtr];
+
+        uint32 numberOfStateDiffs = uint32(bytes4(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 4]));
+        calldataPtr += 4;
+
+        bytes calldata stateDiffs = _totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr +
+            (numberOfStateDiffs * STATE_DIFF_ENTRY_SIZE)];
+
+        calldataPtr += numberOfStateDiffs * STATE_DIFF_ENTRY_SIZE;
+
+        repackStateDiffs(chainId, numberOfStateDiffs, enumerationIndexSize, stateDiffs, compressedStateDiffs);
+    }
+    function bytesLength(uint256 value) internal returns (uint8 length){
+        while(value>0){
+            length += 1;
+            value = value>>8;
+        }
+    }
+    
+    function compressValue(uint256 initialValue, uint256 finalValue) internal returns (bytes memory compressedDiff){
+        uint8 transform = bytesLength(finalValue);
+        uint8 add = bytesLength(finalValue-initialValue);
```

Comment:
this will panic if `finalValue < initialValue`. please use `unchecked`

Reply:
fixed